lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart

```diff
@@ -7,25 +7,40 @@ import 'dart:js_interop';
 
 import 'package:ui/src/engine.dart';
 
-abstract class SkwasmObjectWrapper<T extends NativeType> {
-  Pointer<T> get handle;
+class SkwasmObjectWrapper<T extends NativeType> {
+  SkwasmObjectWrapper(this.handle, this.registry) {
+    registry.register(this);
+  }
+  final SkwasmFinalizationRegistry<T> registry;
+  final Pointer<T> handle;
+  bool _isDisposed = false;
+
+  void dispose() {
+    assert(!_isDisposed);
+    registry.evict(this);
+    _isDisposed = true;
+  }
+
+  bool get debugDisposed => _isDisposed;
 }
 
 typedef DisposeFunction<T extends NativeType> = void Function(Pointer<T>);
 
 class SkwasmFinalizationRegistry<T extends NativeType> {
-  SkwasmFinalizationRegistry(DisposeFunction<T> dispose)
+  SkwasmFinalizationRegistry(this.dispose)
     : registry = DomFinalizationRegistry(((JSNumber address) =>
       dispose(Pointer<T>.fromAddress(address.toDart.toInt()))
     ).toJS);
 
   final DomFinalizationRegistry registry;
+  final DisposeFunction<T> dispose;
```

Comment:
nit: could these fields be private?

Reply:
One unusual aspect of our codebase is that everything under `src/engine/` is automatically private, because we rewrite the code into `dart:_engine`. So it's OK to leave things public if it helps with testing and such.

Reply:
defaulting to private makes it easy to understand what's used. It also helps find when things are no longer used.